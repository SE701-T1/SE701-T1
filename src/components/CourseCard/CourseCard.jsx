import React, { useState } from 'react';
import { styled } from '@mui/material/styles';
import Box from '@mui/material/Box';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import Typography from '@mui/material/Typography';
import { CardActionArea, Grid } from '@mui/material';
import PersonIcon from '@mui/icons-material/Person';
import GroupsIcon from '@mui/icons-material/Groups';
import CheckCircleOutlinedIcon from '@mui/icons-material/CheckCircleOutlined';
import PropTypes from 'prop-types';

import styles from './CourseCard.module.css';

const SelectedCard = styled(Card)(() => ({
  boxShadow: '0px 0px 0px 4.5px #8896ac inset',
  borderRadius: '15px',

  position: 'relative',
  '&::before,&::after': {
    content: "''",
    position: 'absolute',
    top: 0,
    right: 0,
    borderColor: 'transparent',
    borderStyle: 'solid',
  },
  '&::before': {
    borderWidth: '4px',
    borderRightColor: 'transparent',
    borderTopColor: 'transparent',
  },
  '&::after': {
    borderRadius: '0%',
    borderWidth: '30px',
    borderRightColor: '#8896ac',
    borderTopColor: '#8896ac',
  },
}));

const StyledCard = styled(Card)(() => ({
  borderRadius: '15px',
}));

function CourseCard({ courseName, semesterNumber, numbOfStudents, numOfBuddies }) {
  const [selected, setSelected] = useState(false);

  const cardContent = (
    <CardContent className={styles.cardContent}>
      <Grid container direction="row" spacing={3}>
        <Grid item>
          <Typography variant="h4" fontWeight={550} className={styles.title}>
            {courseName}
          </Typography>
          <Typography className={styles.semesterText} fontSize={14}>
            {semesterNumber}
          </Typography>
        </Grid>
      </Grid>

      <Grid
        container
        direction="row"
        alignItems="baseline"
        spacing={1}
        className={styles.memberNBuddiesText}>
        <Grid item>
          <Grid container direction="column">
            <Grid item>
              <PersonIcon className={styles.icons} />
            </Grid>
            <Grid item>
              <GroupsIcon className={styles.icons} />
            </Grid>
          </Grid>
        </Grid>
        <Grid item>
          <Grid container direction="column" alignItems="flex-end">
            <Grid item>
              <Typography variant="subtitle2" className={styles.text} align="right">
                <strong>{numbOfStudents}</strong>
              </Typography>
            </Grid>
            <Grid item>
              <Typography variant="subtitle2" className={styles.text} align="right">
                <strong>{numOfBuddies}</strong>
              </Typography>
            </Grid>
          </Grid>
        </Grid>
        <Grid item>
          <Grid container direction="column">
            <Grid item>
              <Typography variant="subtitle2" className={styles.text}>
                <strong>Members</strong>
              </Typography>
            </Grid>
            <Grid item>
              <Typography variant="subtitle2" className={styles.text}>
                <strong>Buddies</strong>
              </Typography>
            </Grid>
          </Grid>
        </Grid>
      </Grid>
    </CardContent>
  );

  return (
    <Box className={styles.root}>
      {selected ? (
        <Box className={styles.root} onClick={() => setSelected(!selected)}>
          <SelectedCard className={styles.card}>
            <CardActionArea>{cardContent}</CardActionArea>
          </SelectedCard>
          <CheckCircleOutlinedIcon data-testid="tickIcon" className={styles.tickIcon} />
        </Box>
      ) : (
        <StyledCard className={styles.card}>
          <CardActionArea onClick={() => setSelected(!selected)}>{cardContent}</CardActionArea>
        </StyledCard>
      )}
    </Box>
  );
}

CourseCard.propTypes = {
  courseName: PropTypes.string.isRequired,
  semesterNumber: PropTypes.string.isRequired,
  numbOfStudents: PropTypes.number.isRequired,
  numOfBuddies: PropTypes.number.isRequired,
};

export default CourseCard;
