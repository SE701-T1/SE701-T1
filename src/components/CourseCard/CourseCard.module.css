.root {
  max-width: 280px;
  position: relative;
}

.cardContent {
  padding-top: 20px;

  margin-bottom: 55px;
}

.card {
  position: relative;
}

.icons {
  display: inline;
  color: #666666;
  vertical-align: middle;
}

.semesterText {
  /* font-family: 'inner'; */
  color: #8896ac;
}

.memberNBuddiesText {
  padding-top: 15px;
}

.text {
  display: inline;

  vertical-align: text-top;
}

.tickIcon {
  color: white;
  top: 9px;
  right: 6px;
  position: absolute;
  display: block;
}
