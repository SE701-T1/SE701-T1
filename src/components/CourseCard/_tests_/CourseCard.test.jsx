/* eslint-disable testing-library/await-async-query */
import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import CourseCard from '../CourseCard';

describe('Course Component test', () => {
  it('should display the data', () => {
    render(
      <CourseCard
        courseName="Softeng 701"
        semesterNumber="2022 sem 1"
        numbOfStudents={20}
        numOfBuddies={0}
      />,
    );

    expect(screen.getByText('Softeng 701')).toBeInTheDocument();
    expect(screen.getByText('2022 sem 1')).toBeInTheDocument();
    expect(screen.getByText('20')).toBeInTheDocument();
    expect(screen.getByText('0')).toBeInTheDocument();
  });
  it('should display border', () => {
    render(
      <CourseCard
        courseName="Softeng 701"
        semesterNumber="1"
        numbOfStudents={20}
        numOfBuddies={0}
      />,
    );

    // eslint-disable-next-line testing-library/no-promise-in-fire-event
    userEvent.click(screen.getByRole('button'));
    expect(screen.getByTestId('tickIcon'));
  });
  it('produces the correct initial snapshot', () => {
    const tree = renderer
      .create(
        <CourseCard
          courseName="Softeng 701"
          semesterNumber="1"
          numbOfStudents={20}
          numOfBuddies={0}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
