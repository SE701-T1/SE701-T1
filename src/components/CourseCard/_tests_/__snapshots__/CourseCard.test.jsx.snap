// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course Component test produces the correct initial snapshot 1`] = `
<div
  className="root MuiBox-root css-0"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root card css-4qowu-MuiPaper-root-MuiCard-root"
  >
    <button
      className="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <div
        className="MuiCardContent-root cardContent css-46bh2p-MuiCardContent-root"
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <h5
              className="MuiTypography-root MuiTypography-h5 title css-z49yrg-MuiTypography-root"
            >
              Softeng 701
            </h5>
            <p
              className="MuiTypography-root MuiTypography-body1 semesterText css-1d9ye53-MuiTypography-root"
            >
              1
            </p>
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 memberNBuddiesText css-17oy7hn-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-1vpwcmr-MuiGrid-root"
            >
              <div
                className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icons css-i4bv87-MuiSvgIcon-root"
                  data-testid="PersonIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                  />
                </svg>
              </div>
              <div
                className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icons css-i4bv87-MuiSvgIcon-root"
                  data-testid="GroupsIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12.75c1.63 0 3.07.39 4.24.9 1.08.48 1.76 1.56 1.76 2.73V18H6v-1.61c0-1.18.68-2.26 1.76-2.73 1.17-.52 2.61-.91 4.24-.91zM4 13c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm1.13 1.1c-.37-.06-.74-.1-1.13-.1-.99 0-1.93.21-2.78.58C.48 14.9 0 15.62 0 16.43V18h4.5v-1.61c0-.83.23-1.61.63-2.29zM20 13c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm4 3.43c0-.81-.48-1.53-1.22-1.85-.85-.37-1.79-.58-2.78-.58-.39 0-.76.04-1.13.1.4.68.63 1.46.63 2.29V18H24v-1.57zM12 6c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-1v7j7al-MuiGrid-root"
            >
              <div
                className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <h6
                  className="MuiTypography-root MuiTypography-subtitle2 MuiTypography-alignRight text css-h2l1uj-MuiTypography-root"
                >
                  <strong>
                    20
                  </strong>
                </h6>
              </div>
              <div
                className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <h6
                  className="MuiTypography-root MuiTypography-subtitle2 MuiTypography-alignRight text css-h2l1uj-MuiTypography-root"
                >
                  <strong>
                    0
                  </strong>
                </h6>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-1vpwcmr-MuiGrid-root"
            >
              <div
                className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <h6
                  className="MuiTypography-root MuiTypography-subtitle2 text css-ati914-MuiTypography-root"
                >
                  <strong>
                    Members
                  </strong>
                </h6>
              </div>
              <div
                className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <h6
                  className="MuiTypography-root MuiTypography-subtitle2 text css-ati914-MuiTypography-root"
                >
                  <strong>
                    Buddies
                  </strong>
                </h6>
              </div>
            </div>
          </div>
        </div>
      </div>
      <span
        className="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
      />
    </button>
  </div>
</div>
`;
